### PR TITLE
Response Compression updates (+Brotli)

### LINKED ISSUE #7715: Brotli compression
Sonarwhal.com is telling me to use Brotli and/or Zopfli compression. It'd be great to have an official example of Brotil and Zopfli if available. I've found one here: https://docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-2.1&tabs=aspnetcore2x.

https://sonarwhal.com/docs/user-guide/rules/rule-http-compression/

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9795b92f-db24-8f1f-88c0-c36666d3526f
* Version Independent ID: 2d56d9b6-3d7b-0204-6c08-8e584fccecc2
* Content: [Response Compression Middleware for ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-2.1&tabs=aspnetcore2x)
* Content Source: [aspnetcore/performance/response-compression.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/performance/response-compression.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #7715 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-1.1&branch=pr-en-us-8560)

* A number of topic UE updates are provided.
* Brotli content is added to the topic.
* I use inline code examples over snapshot files because this will only in need of a quick sample update + some sample link changes when 2.2 RTMs, and that's right around the corner.
* Opened **Add Brotli compression to the sample app** #8559 to deal with sample app update later.

cc: @khellang ... do u want to take a look at the Brotli bits before I call for engineering to take a look?

### CODE REVIEW

File: aspnetcore/performance/response-compression.md [hunk 1]
```diff
@@ -36,16 +36,33 @@ Usually, any response not natively compressed can benefit from response compress
...
-| `compress`                      | No                   | UNIX "compress" data format                                 |
-| `deflate`                       | No                   | "deflate" compressed data inside the "zlib" data format     |
-| `exi`                           | No                   | W3C Efficient XML Interchange                               |
-| `gzip`                          | Yes (default)        | gzip file format                                            |
+::: moniker range=">= aspnetcore-2.2"
+
+| `Accept-Encoding` header values | Middleware Supported | Description |
+| ------------------------------- | :------------------: | ----------- |
+| `br`                            | Yes (default)        | [Brotli compressed data format](https://tools.ietf.org/html/rfc7932) |
+| `deflate`                       | No                   | [DEFLATE compressed data format](https://www.ietf.org/rfc/rfc1951.txt) |
```
Comment (khellang) [inline aspnetcore/performance/response-compression.md]: Maybe all of these should use the `tools.ietf.org` links? Or at least be consistent?

Comment (guardrex) [reply aspnetcore/performance/response-compression.md]: Thanks ... yes, I'll do that. I was relying too heavily on search results, and top results were inconsistent.

Comment (guardrex) [reply aspnetcore/performance/response-compression.md]: @khellang I found one I could change to `www.ietf.org` but not the other two. I'll see if John has any further feedback about it on review. Thanks again for taking a look.

**Comment by guardrex** [comment]
@JunTaoLuo You reviewed the original response compression topic. Do you want to take a look at these updates (mostly adding Brotli), or can you recommend someone else?

File: aspnetcore/performance/response-compression.md [hunk 2]
```diff
@@ -66,28 +83,44 @@ The headers involved in requesting, sending, caching, and receiving compressed c
...
+* The compression of app responses using Gzip and custom compression providers.
 * How to add a MIME type to the default list of MIME types for compression.
 
 ## Package
 
-::: moniker range="< aspnetcore-2.1"
+::: moniker range=">= aspnetcore-2.1"
 
-To include the middleware in your project, add a reference to the [Microsoft.AspNetCore.ResponseCompression](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCompression/) package. This feature is available for apps that target ASP.NET Core 1.1 or later.
+To include the middleware in a project, add a reference to the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app), which includes the [Microsoft.AspNetCore.ResponseCompression](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCompression/) package.
```
Comment (scottaddie) [inline aspnetcore/performance/response-compression.md]: This sentence should be friendly to those targeting .NET Framework too. I suggest leading with the name of the ResponseCompression package.

File: aspnetcore/performance/response-compression.md [hunk 3]
```diff
@@ -66,28 +83,44 @@ The headers involved in requesting, sending, caching, and receiving compressed c
...
 
 ::: moniker-end
 
-::: moniker range=">= aspnetcore-2.1"
+::: moniker range="= aspnetcore-2.0"
 
-To include the middleware in your project, add a reference to the [Microsoft.AspNetCore.ResponseCompression](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCompression/) package or use the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app) (ASP.NET Core 2.1 or later).
+To include the middleware in a project, add a reference to the [Microsoft.AspNetCore.All metapackage](xref:fundamentals/metapackage), which includes the [Microsoft.AspNetCore.ResponseCompression](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCompression/) package.
```
Comment (scottaddie) [inline aspnetcore/performance/response-compression.md]: Same suggestion as above.

File: aspnetcore/performance/response-compression.md [hunk 4]
```diff
@@ -117,77 +150,248 @@ Submit a request to the sample app with the `Accept-Encoding: gzip` header and o
...
 
 Submit a request to the sample app with the `Accept-Encoding: mycustomcompression` header and observe the response headers. The `Vary` and `Content-Encoding` headers are present on the response. The response body (not shown) isn't compressed by the sample. There isn't a compression implementation in the `CustomCompressionProvider` class of the sample. However, the sample shows where you would implement such a compression algorithm.
 
 ![Fiddler window showing result of a request with the Accept-Encoding header and a value of mycustomcompression. The Vary and Content-Encoding headers are added to the response.](response-compression/_static/request-custom-compression.png)
 
+## MIME types
+
+The middleware specifies a default set of MIME types for compression:
+
+* `text/plain`
```
Comment (scottaddie) [inline aspnetcore/performance/response-compression.md]: Order this list alphabetically for easier scanning.

File: aspnetcore/performance/response-compression.md [hunk 5]
```diff
@@ -214,21 +418,34 @@ When a request is proxied by Nginx, the `Accept-Encoding` header is removed. Thi
...
+If you have an active IIS Dynamic Compression Module configured at the server level that you would like to disable for an app, disable the module with an addition to the *web.config* file. For more information, see [Disabling IIS modules](xref:host-and-deploy/iis/modules#disabling-iis-modules).
 
 ## Troubleshooting
 
-Use a tool like [Fiddler](http://www.telerik.com/fiddler), [Firebug](http://getfirebug.com/), or [Postman](https://www.getpostman.com/), which allow you to set the `Accept-Encoding` request header and study the response headers, size, and body. The Response Compression Middleware compresses responses that meet the following conditions:
+Use a tool like [Fiddler](http://www.telerik.com/fiddler), [Firebug](http://getfirebug.com/), or [Postman](https://www.getpostman.com/), which allow you to set the `Accept-Encoding` request header and study the response headers, size, and body. By default, Response Compression Middleware compresses responses that meet the following conditions:
```
Comment (scottaddie) [inline aspnetcore/performance/response-compression.md]: http --> https

File: aspnetcore/performance/response-compression.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/performance/response-compression.md]: I like leading with the package name we're pushing. We're pushing `App`.

**Pull request merged**